habitat-lab/habitat/tasks/rearrange/actions/actions.py

```diff
@@ -467,3 +467,48 @@ def step(self, ee_pos, **kwargs):
             self._sim.viz_ids["ee_target"] = self._sim.visualize_position(
                 global_pos, self._sim.viz_ids["ee_target"]
             )
+
+
+@registry.register_task_action
+class HumanoidJointAction(RobotAction):
+    def __init__(self, *args, sim: RearrangeSim, **kwargs):
+        super().__init__(*args, sim=sim, **kwargs)
+        self._sim: RearrangeSim = sim
+        self.num_joints = 19
+
+    def reset(self, *args, **kwargs):
+        super().reset()
+
+    @property
+    def action_space(self):
+        num_joints = self.num_joints
+        num_dim_transform = 16
+        # The action space is the number of joints plus 16 for a 4x4 transformtion matrix for the base
+        return spaces.Dict(
+            {
+                "human_joints_trans": spaces.Box(
+                    shape=(4 * (num_joints + num_dim_transform),),
+                    low=-1,
+                    high=1,
+                    dtype=np.float32,
+                )
+            }
+        )
+
+    def step(self, **kwargs):
+        new_pos_transform = kwargs["human_joints_trans"]
+        new_joints = new_pos_transform[:-16]
+        new_pos_transform = new_pos_transform[-16:]
+
+        # When the array is all 0, this indicates we are not setting
+        # the human joint
+        if np.array(new_pos_transform).sum() != 0:
+            vecs = [
+                mn.Vector4(new_pos_transform[i * 4 : (i + 1) * 4])
+                for i in range(4)
+            ]
+            new_transform = mn.Matrix4(*vecs)
+            self.cur_articulated_agent.set_joint_transform(
+                new_joints, new_transform
+            )
+        return self._sim.step(HabitatSimActions.changejoint_action)
```

Comment:
I do not understand what this does. Why is this needed? What does self._sim need to update for the changejoint_action ?

Reply:
I probably misunderstood the code, but don't we need to call `sim.step` on every step() function, since functions calling step expect observations from the environment? I think this makes sense when we call this function via the high level policy.

Reply:
This could cause the simulator to step twice per high-level step. The action should only step the simulator if the action is the last action to execute. See how it's [done in the other actions](https://github.com/facebookresearch/habitat-lab/blob/2ca5b88bf86bb9fef790eebee9dddc08bc0bf4c9/habitat-lab/habitat/tasks/rearrange/actions/actions.py#L403).